### PR TITLE
feat: unordered scanner scans data by time ranges

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR adds the initial implementation for the unordered scanner to support reading rows from time ranges individually.

The scanner groups sources (files and memtables) into ranges and maintains a `RangeMeta` to store the metadata of that range. Each `PartitionRange` refers to a `RangeMeta` in the `StreamContext` so we can use the `PartitionRange::identifier` to locate the `RangeMeta`.

`SeqScan` and `UnorderedScan` have different grouping strategies:
- `SeqScan` merges all overlapping files and memtables to ensure output rows are sorted. So it puts overlapping files and memtables in the same range.
  - If the range only consists of one source, then it can further split row groups of the source into multiple ranges
- `UnorderedScan` can assign each row group to a dedicated range as it doesn't sort and merge duplication rows

The implementations of grouping strategies are similar to `SeqDistributor` and `UnorderedDistributor`.

It also refactors the `scan_partition()` method by wrapping some methods to scan memtables and files.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Release Notes

- **New Features**
  - Introduced a new module for managing partition ranges, enhancing data scanning capabilities.
  - Added methods to handle pruning of memtable and file ranges, improving efficiency during scans.

- **Improvements**
  - Updated return types for several methods to utilize `BTreeMap`, enhancing data organization and access.
  - Enhanced internal state management with the addition of `num_ranges` to track range counts.

- **Bug Fixes**
  - Removed unnecessary code attributes to streamline the codebase.

- **Documentation**
  - Added comments for future development tasks to improve code maintainability.

### CODE REVIEW

**Comment by evenyag** [comment]
@coderabbitai summary

File: src/mito2/src/read/range.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    }
+    if let Some(range) = range_in_progress {
+        exclusive_ranges.push(range);
+    }
+
+    exclusive_ranges
+}
+
+/// Splits the range into multiple smaller ranges.
+/// It assumes the input `ranges` list is created by [group_ranges_for_merge_mode()].
```
Comment (waynexia) [inline src/mito2/src/read/range.rs:251]: What's `group_ranges_for_merge_mode`?

File: src/mito2/src/read/unordered_scan.rs [hunk 2]
```diff
@@ -127,81 +120,97 @@ impl UnorderedScan {
...
-            metrics.build_reader_cost = build_reader_start.elapsed();
-
-            let query_start = stream_ctx.query_start;
             let cache = stream_ctx.input.cache_manager.as_deref();
-            // Scans memtables first.
-            for mut source in memtable_sources {
-                while let Some(batch) = Self::fetch_from_source(&mut source, mapper, cache, None, &mut metrics).await? {
+            stream_ctx.build_mem_ranges(index, ranges);
+            metrics.num_mem_ranges += ranges.len();
+            for range in &*ranges {
```
Comment (waynexia) [inline src/mito2/src/read/unordered_scan.rs:164]:
```suggestion
            for range in ranges {
```

File: src/mito2/src/read/range.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    // Sorts ranges by time range (start asc, end desc).
+    ranges.sort_unstable_by(|a, b| {
+        let l = a.time_range;
+        let r = b.time_range;
+        l.0.cmp(&r.0).then_with(|| r.1.cmp(&l.1))
+    });
+    let mut range_in_progress = None;
+    // Parts with exclusive time ranges.
+    let mut exclusive_ranges = Vec::with_capacity(ranges.len());
+    for range in ranges {
```
Comment (v0y4g3r) [inline src/mito2/src/read/range.rs:228]: Looks like we have similar code here https://github.com/GreptimeTeam/greptimedb/blob/02df58c92a89bbd7f10315326e1b98139fa153b8/src/mito2/src/read/seq_scan.rs#L625-L648

**Review by waynexia — APPROVED** [review]

File: src/mito2/src/read/seq_scan.rs [hunk 4]
```diff
@@ -68,11 +68,10 @@ impl SeqScan {
     pub(crate) fn new(input: ScanInput) -> Self {
         let parallelism = input.parallelism.parallelism.max(1);
         let mut properties = ScannerProperties::default()
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/read/seq_scan.rs:70]: We can't remove it until we refactor the SeqScan to use RangeMeta.

File: src/mito2/src/read/range.rs (hunk 3 above)
Comment (evenyag) [reply src/mito2/src/read/range.rs:228]: I'll refactor SeqScan in the next PR. After that, we can remove this.

File: src/mito2/src/read/range.rs [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    /// Returns true if the time range of given `meta` overlaps with the time range of this meta.
+    pub(crate) fn overlaps(&self, meta: &RangeMeta) -> bool {
+        overlaps(&self.time_range, &meta.time_range)
+    }
+
+    /// Merges given `meta` to this meta.
+    /// It assumes that the time ranges overlap and they don't have the same file or memtable index.
+    pub(crate) fn merge(&mut self, mut other: RangeMeta) {
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/read/range.rs:77]: resolved in https://github.com/GreptimeTeam/greptimedb/pull/4757/commits/3d419fc804fb9ac7ee0d979d51944f958fda687b

**Review by WenyXu — APPROVED** [review]
LGTM

**Pull request merged**